p2p/stream/protocols/sync/protocol.go

```diff
@@ -0,0 +1,260 @@
+package sync
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/harmony-one/harmony/consensus/engine"
+	nodeconfig "github.com/harmony-one/harmony/internal/configs/node"
+	shardingconfig "github.com/harmony-one/harmony/internal/configs/sharding"
+	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/harmony-one/harmony/p2p/discovery"
+	"github.com/harmony-one/harmony/p2p/stream/common/ratelimiter"
+	"github.com/harmony-one/harmony/p2p/stream/common/requestmanager"
+	"github.com/harmony-one/harmony/p2p/stream/common/streammanager"
+	sttypes "github.com/harmony-one/harmony/p2p/stream/types"
+	"github.com/hashicorp/go-version"
+	libp2p_host "github.com/libp2p/go-libp2p-core/host"
+	libp2p_network "github.com/libp2p/go-libp2p-core/network"
+	"github.com/rs/zerolog"
+)
+
+const (
+	// serviceSpecifier is the specifier for the service.
+	serviceSpecifier = "sync"
+)
+
+var (
+	version100, _ = version.NewVersion("1.0.0")
+
+	// MyVersion is the version of sync protocol of the local node
+	MyVersion = version100
+
+	// MinVersion is the minimum version for matching function
+	MinVersion = version100
+)
+
+type (
+	// Protocol is the protocol for sync streaming
+	Protocol struct {
+		chain    engine.ChainReader            // provide SYNC data
+		schedule shardingconfig.Schedule       // provide schedule information
+		rl       ratelimiter.RateLimiter       // limit the incoming request rate
+		sm       streammanager.StreamManager   // stream management
+		rm       requestmanager.RequestManager // deliver the response from stream
+		disc     discovery.Discovery
+
+		config Config
+		logger zerolog.Logger
+
+		ctx    context.Context
+		cancel func()
+		closeC chan struct{}
+	}
+
+	// Config is the sync protocol config
+	Config struct {
+		Chain     engine.ChainReader
+		Host      libp2p_host.Host
+		Discovery discovery.Discovery
+		ShardID   nodeconfig.ShardID
+		Network   nodeconfig.NetworkType
+
+		// stream manager config
+		SmSoftLowCap int
+		SmHardLowCap int
+		SmHiCap      int
+		DiscBatch    int
+	}
+)
+
+// NewProtocol creates a new sync protocol
+func NewProtocol(config Config) *Protocol {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	sp := &Protocol{
+		chain:  config.Chain,
+		disc:   config.Discovery,
+		config: config,
+		ctx:    ctx,
+		cancel: cancel,
+		closeC: make(chan struct{}),
+	}
+	smConfig := streammanager.Config{
+		SoftLoCap: config.SmSoftLowCap,
+		HardLoCap: config.SmHardLowCap,
+		HiCap:     config.SmHiCap,
+		DiscBatch: config.DiscBatch,
+	}
+	sp.sm = streammanager.NewStreamManager(sp.ProtoID(), config.Host, config.Discovery,
+		sp.HandleStream, smConfig)
+
+	sp.rl = ratelimiter.NewRateLimiter(sp.sm, 50, 10)
```

Comment:
comment on the 50, 10? or make them constants?

Reply:
Let me make them constants. Thanks